### ISSUE TITLE
nvm-windows uses hardcoded urls to download node & npm

### ISSUE THREAD

**Issue by marty-buly** [coreybutler/nvm-windows#83]
In our environment, we have proxy and malware filter enabled for all network traffic outside of our corporate network. So in order to have access to all version of nodejs & npm (which some versions contain binaries blocked by our malware filter), we have local mirrors of those two dist sites.

It would be very nice to have separate settings for nodejs & npm download urls. This way people in strictly managed corporate environments, like us, would be able to use nvm-windows as well.

Would it be possible to include this feature in the upcoming releases of nvm-windows?

**Comment by coreybutler** [comment]
Thanks for reporting this. This is something I'll consider supporting in the next version. I'll need a little time to think through the implementation. My biggest concern is distribution.

In your environment, how are you attempting to install NVM for Windows? Would users have local administrative rights? Does the installer get repackaged and deployed (perhaps through an AD environment)? Ultimately, I'll probably just make it all adjustable via the CLI, but I might be able to put an option in the installer to set it up initially.

**Comment by marty-buly** [comment]
Hi,
we don't have local admin rights. Usually all the software is auto installed for us centrally. So the option with CLI is more suitable.
I had a look at it and tried to work on fix. I was just enriching settings.xml with two url options, and then having in the code your hardcoded defaults. So users would be able to change it when needed. Could this be sufficient? It should work for us. Haven't tested it yet, but that would be the logic behind.

**Comment by damnhandy** [comment]
+1 My organization has similar limitations: no admin rights and we need to make the software available internally. Right now, we are using JFrog Artifactory to redistribute the Node.js MSI. We are using the [Scoop](https://github.com/lukesampson/scoop) package manager to install Node, which gets around the need to have Admin rights. Basically, the only thing we are doing is changing the URL where we retrieve node, but it's the same MSI but it is installed to `~\AppData\Local\scoop\apps\nodejs`

**Comment by coreybutler** [comment]
@marty-buly & @damnhandy - Do your environments also maintain a list of available releases, or do they mirror the entire node ecosystem? From a usability perspective, I can see people getting frustrated if  running `nvm list available` doesn't sync with what is truly available.

**Comment by damnhandy** [comment]
Yes. We create a full mirror of https://nodejs.org/dist/ so the only thing that would change is prefix. We're using an Artifactory Remote repo that proxies https://nodejs.org/dist/.

**Comment by damnhandy** [comment]
To be more specific, I'd very happy if we would do something like the [Maven Frontend Plugin](https://github.com/eirslett/frontend-maven-plugin#installing-node-and-npm) and the [Gradle Node Plugin](https://github.com/srs/gradle-node-plugin#configuring-the-plugin), which both allow you to specify a `downloadRoot` to get the node binaries from. It makes the assumption that the `downloadRoot` is 100% compatible with the Node.JS distribution site. Both of these plugins work great against our Artifactory instance.

**Comment by marty-buly** [comment]
I completely agree with @damnhandy. We have the same setup and usage case.

**Comment by coreybutler** [comment]
It's been a long time coming, but thanks to a PR by @lijie371, node mirrors are now in the master branch. I'll close this when the installer has been updated.

**Comment by coreybutler** [comment]
Node mirrors are now released in 1.1.1.
